Added doc explaining how we did one click deployment using AWS

rakaar opened the pull request:
Added doc explaining how we did one-click deployment using AWS. This might be useful when someone wants to know in brief how things are set up in AWS.

- [X] Adding steps to get task def JSON and edit Tast Def 
- [x] Move to Misc folder
- [x] Add about AWS VPC and Subnets

hargup commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+- **<u>Elastic Container Service(ECS) Fargate</u>**: 
+
+  Once the latest image is stored in the ECR, they need to be deployed. A service by name **ECS Fargate** handles this job of deploying the docker image. ECS Fargate is configured using a JSON file present in the `deploy` folder. The JSON is obtained from the AWS Console(Instructions to obtain JSON from AWS console are pinned in Felvin Slack). 
```
Consider all slack conversation transient. It will go away. Please add another doc for these instructions.

rakaar replied:
Added the doc and referenced it!

hargup reviewed (changes requested):
@rakaar Thanks for writing the doc. It explains the concept behind one click deployment using AWS. Though it isn't a recipe. A reciepe gives step by step instructions, if you follow the steps as given, you should be get the thing done without you knowing any other details.

In this case, someone like @chirag828049 should be able to setup one click deployment for https://github.com/felvin-search/project-ias just by following the steps give in the recipe.

Ideally the recipe doc should look something like this:
```
How to setup one-click deployment?

Step 1. <>
Step 2. <>
Step 3. <>
Here, your one-click deployment is ready.
```
The description of the terms should go to [Glossary](https://github.com/felvin-search/docs/pull/10) document, and explaination of other concepts should go at the end of the document.

mukul-mehta commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+2. **Task definition:** Details of the machine where Image is deployed. Things like Memory size, CPU units, Network mode are specified. The Network mode is AWS **Virtual Private Cloud(VPC)**. In VPC, we have 2 subnets, where our images are deployed.
+
+3. **Service**: A service is a collection of Tasks. The service can be of Fargate type(like ours) or EC2 type  or External type. We are using Fargate Service because on our behalf it defines resources, manages and runs the resources. The below image explains better.
```
I'm not sure if this should be called a collection of tasks. A service takes a task definition and adds more details (Such as the VPC, subnets, ports, load balancers) and schedules it to be run. AFAIK one service runs just one task definition. The service controls how many running instances do you want (Minimum and maximum instances that should exist at a given time), health checks to be configured, the load balancer to use, and the production port to bind to. 

I think what you want to mention is that a service controls how many tasks will exist. A task is an instance of a service, which is a recipe. A task is an actual container running the task definition we've provided, and exposing the ports mentioned

It would be better to separate task definition from task, to avoid this confusion IMO

rakaar replied:
Thanks, added the suggestions

mukul-mehta commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+   ![](https://imgur.com/XJhTotJ.png)
+
+    Here, we have only one Service, since we have only a single task. Our service has a Load Balancer. And the Load Balancer has a target group(Load Balancer and Target Group will be explained soon)
```
We're using one service, and we've set that service to spawn only 1 task

rakaar replied:
Thanks, added !

mukul-mehta commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+    Here, we have only one Service, since we have only a single task. Our service has a Load Balancer. And the Load Balancer has a target group(Load Balancer and Target Group will be explained soon)
+
+4. **Cluster**: A cluster is a collection of services. We are using only one cluster, as we need only one Fargate Service.
```
_We are using only one cluster, as we need only one Fargate Service_

This doesn't feel right. A cluster is a logical collections of all services. Even if we had more than one service, we would use a single cluster. We have one service and one task definition in this specific case, but we could always add more services to this cluster

rakaar replied:
Ya, u are right. Changed. thanks

mukul-mehta commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+  After every release, Fargate spins up a new instance, and deploys the latest image. The problem here is, the Public IP address changes after every release. But to configure Domain name, we need to have a single domain address. We can't keep changing Domain Name configuration after every release.
+
+  As per [AWS Docs](https://aws.amazon.com/premiumsupport/knowledge-center/ecs-fargate-static-elastic-ip-address/), a Load Balancer is recommended to solve the above issue. The steps mentioned in the doc were followed to setup a Load Balancer. We have mentioned in our service definition that our image should be deployed in a VPC which contains 2 subnets. The Load balancer sits in front of 2 Subne ...
```
I think _The Load balancer sits in front of 2 Subnets_ is vague. The load balancer balances load across tasks of our service, which we've configured to be run in these 2 subnets. So the load balancer will register targets that are present in the subnets we've configured the tasks to use

mukul-mehta replied:
The load balancer forwards requests made on a certain port to a target group. The load balancer is not aware of the IPs or instances, it only forwards to a target group, which handles sending it to the actual IP or Instance

mukul-mehta replied:
You could also add that the load balancer has a bunch of listener ports. For example, we have a listener 80 which listens for requests sent to port 80, and then knows where to forward all port 80 requests

rakaar replied:
Resolved after conversation 
Thanks

mukul-mehta commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+  As per [AWS Docs](https://aws.amazon.com/premiumsupport/knowledge-center/ecs-fargate-static-elastic-ip-address/), a Load Balancer is recommended to solve the above issue. The steps mentioned in the doc were followed to setup a Load Balancer. We have mentioned in our service definition that our image should be deployed in a VPC which contains 2 subnets. The Load balancer sits in front of 2 Subne ...
+
+  To tell the Load Balancer where to forward the request, we need to create a `Target Group`.  A Target group contains the Private IP address of the instance to which request should be forwarded. The Target group is specified to the Load Balancer.
```
Mention that you don't need to manually register targets in the target group. ECS will pick up the Fargate instances, and register them. This is because we've specified the container port to load balance in the service

rakaar replied:
thanks, added!

rakaar commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+# Recipe for One-Click Deployment
+
+This document contains a brief explanation on how we have setup one-click deployment. By one-click deployment, we mean by just making a release on Github, the changes are deployed without any human action. Many of the items used from the AWS menu are explained below.
```
Done, nice point!
Thanks!

rakaar commented on troubleshooting-and-how-tos/deployment.md:
```diff
@@ -0,0 +1,50 @@
+
+1. **Container Defintion**: Details of the Docker container. We need to specify the docker Image URI, which is already existing in ECR.
+
+2. **Task definition:** Details of the machine where Image is deployed. Things like Memory size, CPU units, Network mode are specified. The Network mode is AWS **Virtual Private Cloud(VPC)**. In VPC, we have 2 subnets, where our images are deployed.
```
Yes, you are right, thanks for the suggestions. I changed them as told.

mukul-mehta reviewed (approved):
After a discussion with @rakaar, made changes where required. Looks awesome now!

rakaar commented:
Thanks a lot for the help @mukul-mehta

rakaar commented:
@hargup 
I actually meant to provide an overview of how things are happening in AWS. Providing a step by step process is a practically impossible thing to be done in the case of AWS. Some of the steps are complex and subtle.  The doc will help for someone wants to know how we have setup deployment using AWS.

I have changed the title. Also moved the doc to `misc` folder.

OrkoHunter reviewed (approved):
Looks good to me

OrkoHunter commented:
We'll do some re-ordering and refactors as well as soon as the new docs site lands.
